jerry-core/ecma/builtin-objects/ecma-builtin-regexp.c

```diff
@@ -90,36 +73,77 @@ ecma_builtin_regexp_dispatch_construct (const ecma_value_t *arguments_list_p, /*
     return regexp_value;
   }
 
-  bool is_regexp = regexp_value == ECMA_VALUE_TRUE;
+  bool pattern_is_regexp = regexp_value == ECMA_VALUE_TRUE;
+  re_compiled_code_t *bc_p = NULL;
+
+  if (new_target_p == NULL)
+  {
+    new_target_p = ecma_builtin_get (ECMA_BUILTIN_ID_REGEXP);
+
+    if (pattern_is_regexp && ecma_is_value_undefined (flags_value))
+    {
+      ecma_object_t *pattern_obj_p = ecma_get_object_from_value (pattern_value);
+
+      ecma_value_t pattern_constructor = ecma_op_object_get_by_magic_id (pattern_obj_p, LIT_MAGIC_STRING_CONSTRUCTOR);
+
+      if (ECMA_IS_VALUE_ERROR (pattern_constructor))
+      {
+        return pattern_constructor;
+      }
+
+      bool is_same = ecma_op_same_value (ecma_make_object_value (new_target_p), pattern_constructor);
+      ecma_free_value (pattern_constructor);
+
+      if (is_same)
+      {
+        return ecma_copy_value (pattern_value);
+      }
+    }
+  }
+
+  if (ecma_object_is_regexp_object (pattern_value))
+  {
+    ecma_extended_object_t *pattern_obj_p = (ecma_extended_object_t *) ecma_get_object_from_value (pattern_value);
+
+    bc_p = ECMA_GET_INTERNAL_VALUE_POINTER (re_compiled_code_t,
+                                            pattern_obj_p->u.class_prop.u.value);
+
+    if (ecma_is_value_undefined (flags_value))
+    {
+      create_regexp_from_bc = true;
+    }
+    else
+    {
+      create_regexp_from_bc = false;
+      pattern_value = bc_p->source;
+    }
```

Comment:
This could be simplified like:
```c
create_regexp_from_bc = ecma_is_value_undefined (flags_value);

if (!create_regexp_from_bc) {
  ecma_extended_object_t *pattern_obj_p = (ecma_extended_object_t *) ecma_get_object_from_value (pattern_value);
  bc_p = ECMA_GET_INTERNAL_VALUE_POINTER (re_compiled_code_t,
                                          pattern_obj_p->u.class_prop.u.value);

  pattern_value = bc_p->source;
}
```